File: Backend/src/main/java/ma/hafidelmoudden/backend/services/RemboursementService.java
package ma.hafidelmoudden.backend.services;

import ma.hafidelmoudden.backend.dtos.RemboursementDTO;
import ma.hafidelmoudden.backend.enums.TypeRemboursement;

import java.util.List;

public interface RemboursementService {
    RemboursementDTO saveRemboursement(RemboursementDTO remboursementDTO);
    RemboursementDTO getRemboursementById(Long id);
    List<RemboursementDTO> getAllRemboursements();
    RemboursementDTO updateRemboursement(RemboursementDTO remboursementDTO);
    void deleteRemboursement(Long id);
    List<RemboursementDTO> getRemboursementsByCreditId(Long creditId);
    List<RemboursementDTO> getRemboursementsByType(TypeRemboursement type);
    Double getTotalRemboursementsByCreditId(Long creditId);
    Double getRemainingAmountForCredit(Long creditId);
} 

File: Frontend/src/app/services/remboursement.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Remboursement, TypeRemboursement } from '../models/remboursement.model';
import { environment } from '../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class RemboursementService {
  private apiUrl = environment.apiUrl + '/remboursements';

  constructor(private http: HttpClient) { }

  getAllRemboursements(): Observable<Remboursement[]> {
    return this.http.get<Remboursement[]>(this.apiUrl);
  }

  getRemboursementById(id: number): Observable<Remboursement> {
    return this.http.get<Remboursement>(`${this.apiUrl}/${id}`);
  }

  getRemboursementsByCreditId(creditId: number): Observable<Remboursement[]> {
    return this.http.get<Remboursement[]>(`${this.apiUrl}/credit/${creditId}`);
  }

  getRemboursementsByType(type: TypeRemboursement): Observable<Remboursement[]> {
    return this.http.get<Remboursement[]>(`${this.apiUrl}/type/${type}`);
  }

  createRemboursement(remboursement: Remboursement): Observable<Remboursement> {
    return this.http.post<Remboursement>(this.apiUrl, remboursement);
  }

  updateRemboursement(remboursement: Remboursement): Observable<Remboursement> {
    return this.http.put<Remboursement>(`${this.apiUrl}/${remboursement.id}`, remboursement);
  }

  deleteRemboursement(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  getTotalRemboursementsByCreditId(creditId: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/credit/${creditId}/total`);
  }

  getRemainingAmountForCredit(creditId: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/credit/${creditId}/remaining`);
  }
} 